lib/iob/history.js

```diff
@@ -148,12 +148,26 @@ function calcTempTreatments (inputs, zeroTempDuration) {
                 continue;
             }
             var rate = current.rate;
-            var date = current.date;
-            if (i>0 && pumpHistory[i-1].date == date && pumpHistory[i-1]._type == "TempBasalDuration") {
-                var duration = pumpHistory[i-1]['duration (min)'];
-            } else if (i+1<pumpHistory.length && pumpHistory[i+1].date == date && pumpHistory[i+1]._type == "TempBasalDuration") {
-                var duration = pumpHistory[i+1]['duration (min)'];
-            } else { console.error("No duration found for "+rate+" U/hr basal"+date, pumpHistory[i - 1], current, pumpHistory[i + 1]); }
+            var timestamp = current.timestamp;
+	    var duration;
+            if (i>0 && pumpHistory[i-1].timestamp == timestamp && pumpHistory[i-1]._type == "TempBasalDuration") {
+                duration = pumpHistory[i-1]['duration (min)'];
+            } else { 
+		var iter = i+1;
+
+		while (iter < pumpHistory.length && pumpHistory[iter].timestamp == timestamp) {
```

Comment:
Would it be more straightforward to just make this a for loop over the entire array?  Something like `for (var iter=0; iter < pumpHistory.length; iter++) {` maybe?

Reply:
Yeah, can do that. Any idea why it wasn't done like that in the first place?

Reply:
Efficiency.  But if we're checking the adjacent records first, this is just a fallback for a very small percentage of records that don't match those, so I'm not really concerned about performance here.